### PR TITLE
R: 4.0.4 -> 4.1.0

### PR DESCRIPTION
**Pull request by PhDyellow**
###### Motivation for this change
Bump R to latest release.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by PhDyellow** [comment]
Currently failing tests during build.

I don't know enough about compiling R to figure out what is breaking yet.

File: pkgs/applications/science/math/R/default.nix
Comment (collares) [inline pkgs/applications/science/math/R/default.nix:33]:
```suggestion
  ] ++ lib.optionals (!withRecommendedPackages) [
    (fetchpatch {
       name = "fix-tests-without-recommended-packages.patch";
       url = "https://github.com/wch/r-source/commit/7715c67cabe13bb15350cba1a78591bbb76c7bac.patch";
       # this part of the patch reverts something that was committed after R 4.1.0, so ignore it.
       excludes = [ "tests/Pkgs/xDir/pkg/DESCRIPTION" ];
       sha256 = "sha256-iguLndCIuKuCxVCi/4NSu+9RzBx5JyeHx3K6IhpYshQ=";
    })
    (fetchpatch {
      name = "use-codetools-conditionally.patch";
      url = "https://github.com/wch/r-source/commit/7543c28b931db386bb254e58995973493f88e30d.patch";
      sha256 = "sha256-+yHXB5AItFyQjSxfogxk72DrSDGiBh7OiLYFxou6Xlk=";
    })   
  ];
```

**Review by collares — CHANGES_REQUESTED** [review]
Many thanks for the update! Sorry for taking so long to get to it :( I added a couple of upstream patches that should help with the compilation errors. I will run nixpkgs-review with this change applied to see how it fares.

**Comment by PhDyellow** [comment]
Thanks @collares, I've added the patch to the PR, but I don't have time right now to test it. If it works for you, I'll convert this PR from draft to full PR.

File: pkgs/applications/science/math/R/default.nix
Comment (collares) [inline pkgs/applications/science/math/R/default.nix:45]:
```suggestion
    })
```
I messed up and left a bunch of whitespace here. Sorry! 

In better news, nixpkgs-review is happy with the changes ("13 packages built: R diffoscope ocamlPackages.ocaml-r postgresql11Packages.plr python38Packages.cnvkit python38Packages.rpy2 python39Packages.cnvkit python39Packages.rpy2 rWrapper rstudio rstudioWrapper sage sageWithDoc"), so I think this PR is good to go after the whitespace fix. If you do have the time, squashing the commits would be good.

**Review by collares — APPROVED** [review]

**Comment by collares** [comment]
Result of `nixpkgs-review pr 127798` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>spark</li>
  </ul>
</details>
<details>
  <summary>11 packages built:</summary>
  <ul>
    <li>R</li>
    <li>diffoscope</li>
    <li>postgresqlPackages.plr (postgresql11Packages.plr)</li>
    <li>python38Packages.cnvkit</li>
    <li>python38Packages.rpy2</li>
    <li>python39Packages.cnvkit</li>
    <li>python39Packages.rpy2</li>
    <li>rstudio</li>
    <li>rstudioWrapper</li>
    <li>sage</li>
    <li>sageWithDoc</li>
  </ul>
</details>

**Comment by PhDyellow** [comment]
@collares, is squashing something I should do to the PR branch, or should that be done by whoever merges the PR into Nixpkgs?

**Comment by collares** [comment]
I don't have merge access, so I don't know if mergers typically squash PRs, but I think people usually do it themselves instead of waiting for the committer. If you're comfortable with squashing then feel free to do so! If you're not (or would rather not for any other reason), this PR is fine as is too.

**Review by peti — APPROVED** [review]
Looks good to me. Let's hope CI succeeds.

**Comment by collares** [comment]
This is working fine on x86-{linux,darwin}, but unfortunately fails on aarch64-linux. I tried compiling it on my Android phone (the only aarch64 machine I have access to), and got an error on [this test line](https://github.com/wch/r-source/blob/trunk/tests/Examples/stats-Ex.Rout.save#L11156) with the following log:

```
>    (nlm2 <- update(nlmod, control = list(tol = 8e-8, nDcentral=TRUE), trace=TRUE))
Warning in nls(formula = y ~ Const + A * exp(B * x), algorithm = "default",  :
  No starting values specified for some parameters.
Initializing ‘Const’, ‘A’, ‘B’ to '1.'.
Consider specifying 'start' or using a selfStart model
1017460.306    (4.15e+02): par = (1 1 1)
758164.7503    (2.34e+02): par = (13.42031396 1.961485 0.05947543745)
269506.3540    (3.23e+02): par = (51.75719815 -13.09155955 0.84286077)
68969.21898    (1.03e+02): par = (76.00069849 -1.935226735 1.019085799)
633.3672232    (1.29e+00): par = (100.3761515 8.624648411 5.104490265)
151.4400213    (9.39e+00): par = (100.6344391 4.913490983 0.2849209582)
53.08739799    (7.24e+00): par = (100.6830407 6.899303325 0.4637755078)
1.344478631    (5.97e-01): par = (100.0368306 9.897714142 0.5169294937)
0.9908415908   (1.55e-02): par = (100.0300625 9.9144191 0.5023516843)
0.9906046057   (1.84e-05): par = (100.0288724 9.916224018 0.5025207337)
0.9906046054   (9.94e-08): par = (100.028875 9.916228366 0.50252165)
0.9906046054   (8.70e-08): par = (100.028875 9.916228369 0.5025216506)
0.9906046054   (8.68e-08): par = (100.028875 9.916228369 0.5025216506)
Error in nls(formula = y ~ Const + A * exp(B * x), algorithm = "default",  : 
  step factor 0.000488281 reduced below 'minFactor' of 0.000976562
Calls: update -> update.default -> eval -> eval -> nls
Execution halted
```

I can't be sure ofBorg is seeing the same error, since the logs don't get printed. Either way, I don't know enough about R to diagnose the error above, so any suggestions would be greatly appreciated.

**Comment by PhDyellow** [comment]
Looking at the code, the error, and the comments above the line that breaks on your phone, I think that the line is supposed to test numerical accuracy. It looks like the error means that the model didn't converge fast enough, and the step size went below `minFactor` triggering an error ("your model didn't fit, sorry").

Why is the x86 architecture behaving well, and the aarch64 architecture failing here? I have no idea. Someone from R core team might need to advise us.

**Comment by collares** [comment]
If you know someone from the R core development team that could help us, that would be great! I just checked and Fedora does not bother to run the testsuite on aarch64: https://src.fedoraproject.org/rpms/R/blob/rawhide/f/R.spec#_849

**Comment by PhDyellow** [comment]
I don't know anyone on the core team personally, filing a bug seems like the most practical option.

I think we can either just merge it in anyway, and break R for anyone running R and Nix on aarch64 platforms, or patch out the test that is failing. My guess is that the issue with numerical precision is bad for reproducible research involving precise use of doubles and floats, but fine for almost everything else.

File: pkgs/applications/science/math/R/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/science/math/R/default.nix:45]: Those patches don't break anything when applied all the time, right?

Comment (PhDyellow) [reply pkgs/applications/science/math/R/default.nix:45]: They are reducing test coverage, so I guess they will allow R to build all the time, but funny things might happen during runtime.

**Comment by collares** [comment]
Finally had the time to file https://bugs.r-project.org/show_bug.cgi?id=18165, let's see what happens. Sorry for the delay.

**Comment by collares** [comment]
R 4.1.1 is out! The two patches I suggested are no longer needed. To unblock this, I think we should disable the failing test (preferably just that single line, not the whole `stats-Ex.R` test, maybe only for aarch64) and add a comment linking to the Bugzilla bug report. @peti, what do you think?

**Comment by peti** [comment]
> R 4.1.1 is out! The two patches I suggested are no longer needed. To unblock this PR, I think we should disable the failing test (preferably just that single line, not the whole `stats-Ex.R` test, maybe only for aarch64) and add a comment linking to the above Bugzilla bug report. @peti, what do you think?

Yes, that sounds like a pretty good solution.

**Comment by PhDyellow** [comment]
Great, I can bump R to 4.1.1 and remove the suggested patches. 
However, I'm not in a position to create patches for R that would exclude the failing test. Can anyone else help with that?

File: pkgs/applications/science/math/R/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/science/math/R/default.nix]: Please remove the unused fetchpatch from inputs.

Comment (PhDyellow) [reply pkgs/applications/science/math/R/default.nix]: Done

Comment (SuperSandro2000) [reply pkgs/applications/science/math/R/default.nix]: I think you didn't push this.

**Comment by collares** [comment]
> Great, I can bump R to 4.1.1 and remove the suggested patches.
> However, I'm not in a position to create patches for R that would exclude the failing test. Can anyone else help with that?

This patch should work:

```diff
diff -ur a/src/library/stats/man/nls.Rd b/src/library/stats/man/nls.Rd
--- a/src/library/stats/man/nls.Rd   2021-05-21 19:15:02.000000000 -0300
+++ b/src/library/stats/man/nls.Rd   2021-08-12 12:39:00.094758280 -0300
@@ -287,7 +287,8 @@
 options(digits = 10) # more accuracy for 'trace'
 ## IGNORE_RDIFF_BEGIN
 try(nlm1 <- update(nlmod, control = list(tol = 1e-7))) # where central diff. work here:
-   (nlm2 <- update(nlmod, control = list(tol = 8e-8, nDcentral=TRUE), trace=TRUE))
+ctr2 <- nls.control(tol=8e-8, minFactor=2^-12, nDcentral=TRUE, warnOnly=TRUE)
+   (nlm2 <- update(nlmod, control = ctr2, trace = TRUE))
 ## --> convergence tolerance  4.997e-8 (in 11 iter.)
 ## IGNORE_RDIFF_END
 
```

**Comment by collares** [comment]
This patch should work (I posted a wrong version a minute ago, sorry):

```diff
diff -ur a/src/library/stats/man/nls.Rd b/src/library/stats/man/nls.Rd
--- a/src/library/stats/man/nls.Rd   2021-05-21 19:15:02.000000000 -0300
+++ b/src/library/stats/man/nls.Rd   2021-08-12 12:39:00.094758280 -0300
@@ -287,7 +287,7 @@
 options(digits = 10) # more accuracy for 'trace'
 ## IGNORE_RDIFF_BEGIN
 try(nlm1 <- update(nlmod, control = list(tol = 1e-7))) # where central diff. work here:
-   (nlm2 <- update(nlmod, control = list(tol = 8e-8, nDcentral=TRUE), trace=TRUE))
+   (nlm2 <- update(nlmod, control = list(tol = 8e-8, nDcentral=TRUE, warnOnly=TRUE), trace=TRUE))
 ## --> convergence tolerance  4.997e-8 (in 11 iter.)
 ## IGNORE_RDIFF_END
 
```

File: pkgs/applications/science/math/R/default.nix
Comment (PhDyellow) [reply pkgs/applications/science/math/R/default.nix]: Sorry, I'm not sure what you mean. I can only see two calls two fetchpatch, both of them are removed by d37edcd, and d37edcd was pushed 2 days ago. You did make your original comment before I pushed it. Is this resolved now?

Comment (Artturin) [reply pkgs/applications/science/math/R/default.nix]: Sandro means this https://github.com/PhDyellow/nixpkgs/blob/d37edcd3d5d764b2ac2b9b03835adf744dc596c9/pkgs/applications/science/math/R/default.nix#L4

Comment (PhDyellow) [reply pkgs/applications/science/math/R/default.nix]: Ah, thanks @Artturin!

Comment (PhDyellow) [reply pkgs/applications/science/math/R/default.nix]: Latest commit removes fetchpatch from function inputs.

Comment (PhDyellow) [reply pkgs/applications/science/math/R/default.nix]: Thanks for your feedback @SuperSandro2000

**Comment by PhDyellow** [comment]
Thanks for the patch @collares

File: pkgs/applications/science/math/R/default.nix
Comment (PhDyellow) [reply pkgs/applications/science/math/R/default.nix]: Updating to 4.1.1 removes the need for these patches.

**Review by collares — APPROVED** [review]
Tests pass on aarch64 with the patch disabling that one pesky line, so this should be good to go!

**Comment by 7c6f434c** [comment]
@collares does Sage care about R version?

**Comment by collares** [comment]
@7c6f434c The Sage distribution is still on R 3.6.3 (17 months old), so I doubt it. AFAICT, it depends on R via `rpy2` and, judging by https://rpy2.github.io/doc/latest/html/changes.html, they seem good at maintaining R compatibility. We just need to update `rpy2` every so often.

**Comment by PhDyellow** [comment]
Good to see this merged, thanks particularly to @collares for bringing in the patches.

Bumping the R version got a lot of attention, do people have time and motivation to address some PRs and issues around the r-modules in NixOS?

**Comment by collares** [comment]
@PhDyellow I have never used R and I am not really a maintainer, but if you're mostly fixing build issues then I can at least take a look. If you could have a list of those issues, that would be great.

**Comment by PhDyellow** [comment]
#128735 is a simple one that just fixes build issues on a range of packages.
#129746 fixes R_cache, but #128735 also includes a fix for R_cache among other fixes.
 #127790 makes the packaging of R modules more automatic. It has conflicts now, but I am happy to fix them if I know someone will look at it.
 #93883 is an attempt to generalise R packaging in NixOS beyond CRAN and BioConductor. It currently has some unresolved challenges and isn't ready for merging, but  in the discussion @peti suggested adding other maintainers for R might be helpful.

Of these, #128735 is probably the biggest return on investment.

**Comment by alexvorobiev** [comment]
Does it still produce wrong results when built with MKL (https://github.com/NixOS/nixpkgs/issues/104026)? I had to create an overlay and add settings suggested by Intel (mentioned in that thread).

**Comment by PhDyellow** [comment]
@alexvorobiev, this PR didn't address the MKL issue. I notice you've managed to tweak `R/default.nix` and get it working elsewhere, are you able to open a PR with the changes?

**Pull request merged**
